submariner/globalnet-enhancement2-0.md

```diff
@@ -0,0 +1,298 @@
+    // number of contiguous GlobalIPs for this GlobalEgressIP object.
+    // If unspecified, NumGlobalIPs defaults to 1 and max allowed is restricted to 10.
+    // +optional
+    NumGlobalIPs  int     'json:"numGlobalIPs",omitempty"`
+
+    // Selects the pods whose label match the definition. This is an optional field and
+    // in case it is not set, it results in all the Pods selected from the namespace in which the
+    // CR is created.
+    // +optional
+    PodSelector metav1.LabelSelector `json:"podSelector,omitempty"`
+}
+
+const (
+    // GlobalEgressIPSuccess means that Globalnet was able to successfully allocate the GlobalIP.
+    GlobalEgressIPSuccess GlobalEgressIPStatus = "Success"
+
+    // GlobalEgressIPError means that Globalnet was unable to allocate the GlobalIP.
+    GlobalEgressIPError GlobalEgressIPStatus = "Error"
+)
+
+type EgressIPStatus struct {
+    // Status is one of {"Success", "Error"}
+    Status GlobalEgressIPStatus `json:"status,omitempty"`
+
+    // +optional
+    Message *string `json:"message,omitempty"`
+
+    // The list of GlobalIPs assigned via this GlobalnetEgressIP object.
+    GlobalIPs []string `json:"globalIPs"`
+}
+```
+
+When Submariner Globalnet is deployed in a Cluster, it auto-allocates the configured number of
+`globalIPs` and programs the necessary egress rules on the active Gateway node of the Cluster.
+This is done internally by creating `ClusterGlobalEgressIP` CR object with the name `cluster-default`.
+Users can query the CR to check the allocated GlobalIPs at the Cluster level. If users create
+multiple ClusterGlobalEgressIP CRs that match the Cluster scope, one of the egressIP at the Cluster
+scope will be used.
+
+Only an admin user can create the `ClusterGlobalEgressIP` object and regular users will only be able
+to create `GlobalEgressIP` objects.
+
+### Backward Compatibility
+
+Inorder to support the newer implementation, a combination of ipSets with multiple IPTable chains
+will be used. Also, we will internally use configMaps to maintain the state of allocated GlobalIPs
+and get rid of annotations on the objects. Because of this, backward compatibility cannot be maintained
+with the existing implementation of Globalnet.
+
+### User Stories
+
+1.User has two (or more) clusters with Overlapping CIDRs and requires the Pods in one
+  Cluster be able to discover and talk to exported Services.
+
+  To achieve this use-case, user can simply deploy Submariner Globalnet on all the participating
+  clusters. By default, a globalIP (or a set of globalIPs) will be assigned at the Cluster level
+  which will be used as egressIPs for inter-cluster communication and necessary Services can be
+  exported.
+
+2.In a Globalnet deployment user wants all the Pods in a namespace `ns1` to use a unique globalIP
+  instead of the globalIP assigned at the cluster level.
+
+  To achieve this, one can apply the following CRD in the namespace `ns1`.
+
```

Comment:
What about a GlobalIngressIP for reporting the assigned IPs to ExportedServices? (instead of annotations to ExportedServices?), I was thinking of the issues we had in the past with annotations being removed by the deployment tools sometimes.

Reply:
Yes Miguel, I'd prefer to use GlobalIngressIP CRD with minimal fields instead of annotating the Service/ServiceExport objects based on the past experience.